File: python/tests/run_tests.py
"""
Comprehensive test suite for unicef_api package
Tests all major functionality and saves results to CSV files
"""

import os
import sys
from datetime import datetime

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from unicef_api import get_unicef, list_dataflows, COMMON_INDICATORS, list_vintages
from unicef_api.metadata import sync_metadata, MetadataSync

# Output directory
OUTPUT_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "output")
if not os.path.exists(OUTPUT_DIR):
    os.makedirs(OUTPUT_DIR)

def log(msg):
    print(f"[{datetime.now().strftime('%H:%M:%S')}] {msg}")

def test_list_dataflows():
    """Test listing available dataflows via list_dataflows()"""
    log("Testing list_dataflows()...")
    
    # Use the actual list_dataflows function to ensure consistent output
    df = list_dataflows()
    
    log(f"  Found {len(df)} dataflows")
    
    # Save to CSV - columns should be: id, agency, version, name
    df.to_csv(os.path.join(OUTPUT_DIR, 'test_dataflows.csv'), index=False)
    log(f"  Saved to test_dataflows.csv")
    
    return len(df) > 50  # Should have 60+ dataflows

def test_child_mortality():
    """Test fetching child mortality data (CME_MRY0T4)"""
    log("Testing child mortality (CME_MRY0T4)...")
    
    # Use get_unicef() for consistent column names with R
    df = get_unicef(
        indicator='CME_MRY0T4',
        countries=['USA', 'GBR', 'FRA', 'DEU', 'JPN'],
        start_year=2015,
        end_year=2023
    )
    
    log(f"  Retrieved {len(df)} observations")
    if len(df) > 0:
        log(f"  Countries: {df['iso3'].unique().tolist()}")
        log(f"  Years: {sorted(df['period'].unique().tolist())}")
    
    df.to_csv(os.path.join(OUTPUT_DIR, 'test_mortality.csv'), index=False)
    log(f"  Saved to test_mortality.csv")
    
    return len(df) > 0

def test_stunting():
    """Test fetching stunting data (NT_ANT_HAZ_NE2)"""
    log("Testing stunting (NT_ANT_HAZ_NE2)...")
    
    # Use get_unicef() for consistent column names with R
    df = get_unicef(
        indicator='NT_ANT_HAZ_NE2',
        countries=['IND', 'BGD', 'PAK', 'NPL', 'ETH'],
        start_year=2010,
        end_year=2023
    )
    
    log(f"  Retrieved {len(df)} observations")
    
    df.to_csv(os.path.join(OUTPUT_DIR, 'test_stunting.csv'), index=False)
    log(f"  Saved to test_stunting.csv")
    
    return len(df) > 0

def test_immunization():
    """Test fetching immunization data (IM_DTP3)"""
    log("Testing immunization (IM_DTP3)...")
    
    # Use get_unicef() for consistent column names with R
    df = get_unicef(
        indicator='IM_DTP3',
        countries=['NGA', 'COD', 'BRA', 'IDN', 'MEX'],
        start_year=2015,
        end_year=2023
    )
    
    log(f"  Retrieved {len(df)} observations")
    
    df.to_csv(os.path.join(OUTPUT_DIR, 'test_immunization.csv'), index=False)
    log(f"  Saved to test_immunization.csv")
    
    return len(df) > 0

def test_metadata_sync():
    """Test metadata sync functionality"""
    log("Testing metadata sync...")
    
    cache_dir = os.path.join(OUTPUT_DIR, 'metadata_sync_test')
    sync = MetadataSync(cache_dir=cache_dir)
    results = sync.sync_all(verbose=False)
    
    log(f"  Synced: {results.get('dataflows', 0)} dataflows, {results.get('indicators', 0)} indicators")
    
    # Test vintage listing
    vintages = sync.list_vintages()
    log(f"  Vintages available: {vintages}")
    
    return results.get('dataflows', 0) > 50

def test_multiple_indicators():
    """Test fetching multiple indicators at once"""
    log("Testing multiple indicators...")
    
    indicators = ['CME_MRY0T4', 'CME_MRY0']  # Under-5 and infant mortality
    all_data = []
    
    for ind in indicators:
        try:
            # Use get_unicef() for consistent column names with R
            df = get_unicef(
                indicator=ind,
                countries=['BRA', 'IND', 'CHN'],
                start_year=2020,
                end_year=2023
            )
            all_data.append(df)
            log(f"  {ind}: {len(df)} observations")
        except Exception as e:
            log(f"  {ind}: ERROR - {e}")
    
    if all_data:
        import pandas as pd
        combined = pd.concat(all_data, ignore_index=True)
        combined.to_csv(os.path.join(OUTPUT_DIR, 'test_multiple_indicators.csv'), index=False)
        log(f"  Saved {len(combined)} total observations")
        return True
    return False

def run_all_tests():
    """Run all tests and report results"""
    print("=" * 60)
    print("UNICEF API Python Package Test Suite")
    print(f"Started: {datetime.now()}")
    print("=" * 60)
    
    tests = [
        ("List Dataflows", test_list_dataflows),
        ("Child Mortality", test_child_mortality),
        ("Stunting", test_stunting),
        ("Immunization", test_immunization),
        ("Metadata Sync", test_metadata_sync),
        ("Multiple Indicators", test_multiple_indicators),
    ]
    
    results = []
    for name, test_func in tests:
        try:
            passed = test_func()
            results.append((name, "PASS" if passed else "FAIL", None))
        except Exception as e:
            results.append((name, "ERROR", str(e)))
            log(f"  ERROR: {e}")
    
    print("\n" + "=" * 60)
    print("TEST RESULTS")
    print("=" * 60)
    
    for name, status, error in results:
        icon = "✅" if status == "PASS" else "❌"
        print(f"{icon} {name}: {status}")
        if error:
            print(f"   Error: {error}")
    
    passed = sum(1 for _, s, _ in results if s == "PASS")
    total = len(results)
    print(f"\nTotal: {passed}/{total} tests passed")
    print("=" * 60)
    
    return passed == total

if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)

File: R/tests/run_tests.R
# ============================================================================
# Comprehensive test suite for unicefdata R package
# 
# Test Strategy:
# - OFFLINE tests: Use bundled YAML metadata files (always run, fast)
# - NETWORK tests: Call UNICEF API (skipped in CI, run locally)
#
# Bundled metadata location: R/metadata/current/
#
# Dependencies:
# - Only 'yaml' package needed - CI runs OFFLINE tests that parse bundled
#   metadata files using yaml::read_yaml() without requiring full package
#   dependencies (httr, jsonlite, etc.). Network tests source the full
#   package code only when actually running.
# ============================================================================

# Check environment FIRST
IN_CI <- Sys.getenv("CI") != "" || Sys.getenv("GITHUB_ACTIONS") != ""

# Check actual network connectivity (for true offline detection)
check_network <- function(timeout = 5) {
  old_timeout <- getOption("timeout")
  on.exit(options(timeout = old_timeout), add = TRUE)
  options(timeout = timeout)
  tryCatch({
    # Try to reach UNICEF SDMX API
    con <- url("https://sdmx.data.unicef.org/ws/public/sdmxapi/rest/dataflow", open = "r")
    on.exit(try(close(con), silent = TRUE), add = TRUE)
    TRUE
  }, error = function(e) {
    FALSE
  }, warning = function(w) {
    FALSE
  })
}

# Determine if network tests should run
# Skip if: (1) in CI environment, OR (2) no network connectivity
if (IN_CI) {
  NETWORK_AVAILABLE <- FALSE
  SKIP_REASON <- "CI environment"
} else {
  cat("Checking network connectivity...\n")
  NETWORK_AVAILABLE <- check_network()
  SKIP_REASON <- if (!NETWORK_AVAILABLE) "no network connectivity" else NULL
}

# Determine paths
if (file.exists("R/metadata/current")) {
  METADATA_DIR <- "R/metadata/current"
  OUTPUT_DIR <- "R/tests/output"
  R_DIR <- "R"
} else if (file.exists("../metadata/current")) {
  METADATA_DIR <- "../metadata/current"
  OUTPUT_DIR <- "output"
  R_DIR <- ".."
} else {
  stop("Could not find metadata directory - run from unicefData root")
}

if (!dir.exists(OUTPUT_DIR)) {
  dir.create(OUTPUT_DIR, recursive = TRUE)
}

# Helper function
log_msg <- function(msg) {
  cat(sprintf("[%s] %s\n", format(Sys.time(), "%H:%M:%S"), msg))
}

# ============================================================================
# OFFLINE TESTS - Use bundled YAML metadata (always run, fast, NO network)
# These tests only require the yaml package - no network dependencies
# ============================================================================

test_yaml_dataflows <- function() {
  log_msg("Testing YAML dataflows loading...")
  
  yaml_path <- file.path(METADATA_DIR, "_unicefdata_dataflows.yaml")
  if (!file.exists(yaml_path)) {
    log_msg(sprintf("  SKIP: %s not found", yaml_path))
    return(TRUE)  # Skip gracefully
  }
  
  data <- yaml::read_yaml(yaml_path)
  
  n_dataflows <- length(data$dataflows)
  log_msg(sprintf("  Found %d dataflows in YAML", n_dataflows))
  
  # Verify structure
  first_df <- data$dataflows[[1]]
  has_required <- all(c("id", "name", "agency") %in% names(first_df))
  log_msg(sprintf("  Structure valid: %s", has_required))
  
  # Check specific dataflows exist
  expected <- c("CME", "EDUCATION", "NUTRITION", "IMMUNISATION")
  found <- sum(expected %in% names(data$dataflows))
  log_msg(sprintf("  Key dataflows present: %d/%d", found, length(expected)))
  
  return(n_dataflows >= 50 && has_required && found == length(expected))
}

test_yaml_indicators <- function() {
  log_msg("Testing YAML indicators loading...")
  
  yaml_path <- file.path(METADATA_DIR, "_unicefdata_indicators.yaml")
  if (!file.exists(yaml_path)) {
    log_msg(sprintf("  SKIP: %s not found", yaml_path))
    return(TRUE)
  }
  
  data <- yaml::read_yaml(yaml_path)
  
  n_indicators <- length(data$indicators)
  log_msg(sprintf("  Found %d indicators in YAML", n_indicators))
  
  # Check for expected indicators (some may not be in minimal YAML)
  expected <- c("CME_MRY0T4", "NT_ANT_HAZ_NE2", "IM_DTP3")
  found <- sum(expected %in% names(data$indicators))
  log_msg(sprintf("  Key indicators present: %d/%d", found, length(expected)))
  
  # Minimum threshold: at least 10 indicators (bundled YAML may be minimal)
  return(n_indicators >= 10 && found >= 1)
}

test_yaml_countries <- function() {
  log_msg("Testing YAML countries loading...")
  
  yaml_path <- file.path(METADATA_DIR, "_unicefdata_countries.yaml")
  if (!file.exists(yaml_path)) {
    log_msg(sprintf("  SKIP: %s not found", yaml_path))
    return(TRUE)
  }
  
  data <- yaml::read_yaml(yaml_path)
  
  n_countries <- length(data$countries)
  log_msg(sprintf("  Found %d countries in YAML", n_countries))
  
  # Check for expected countries
  expected <- c("USA", "GBR", "FRA", "DEU", "BRA", "IND", "CHN")
  found <- sum(expected %in% names(data$countries))
  log_msg(sprintf("  Key countries present: %d/%d", found, length(expected)))
  
  return(n_countries >= 150 && found >= 5)
}

test_dataflow_schema_cme <- function() {
  log_msg("Testing CME schema YAML structure...")
  
  schema_path <- file.path(METADATA_DIR, "dataflows", "CME.yaml")
  if (!file.exists(schema_path)) {
    log_msg(sprintf("  SKIP: %s not found", schema_path))
    return(TRUE)
  }
  
  # Read YAML directly (no function call needed)
  schema <- yaml::read_yaml(schema_path)
  
  log_msg(sprintf("  Schema ID: %s", schema$id))
  
  # Extract dimensions
  dimensions <- if (!is.null(schema$dimensions)) {
    sapply(schema$dimensions, function(d) d$id)
  } else {
    character(0)
  }
  
  # Extract attributes  
  attributes <- if (!is.null(schema$attributes)) {
    sapply(schema$attributes, function(a) a$id)
  } else {
    character(0)
  }
  
  log_msg(sprintf("  Dimensions: %d (%s)", length(dimensions), 
                  paste(head(dimensions, 3), collapse = ", ")))
  log_msg(sprintf("  Attributes: %d", length(attributes)))
  
  has_dims <- length(dimensions) > 0
  has_attrs <- length(attributes) > 0
  
  return(has_dims && has_attrs)
}

test_dataflow_schema_education <- function() {
  log_msg("Testing EDUCATION schema YAML structure...")
  
  schema_path <- file.path(METADATA_DIR, "dataflows", "EDUCATION.yaml")
  if (!file.exists(schema_path)) {
    log_msg(sprintf("  SKIP: %s not found", schema_path))
    return(TRUE)
  }
  
  # Read YAML directly
  schema <- yaml::read_yaml(schema_path)
  
  log_msg(sprintf("  Schema ID: %s", schema$id))
  
  dimensions <- if (!is.null(schema$dimensions)) {
    sapply(schema$dimensions, function(d) d$id)
  } else {
    character(0)
  }
  
  log_msg(sprintf("  Dimensions: %s", paste(head(dimensions, 5), collapse = ", ")))
  
  return(schema$id == "EDUCATION" && length(dimensions) > 0)
}

test_schema_files_exist <- function() {
  log_msg("Testing schema files directory...")
  
  schema_dir <- file.path(METADATA_DIR, "dataflows")
  if (!dir.exists(schema_dir)) {
    log_msg(sprintf("  SKIP: %s not found", schema_dir))
    return(TRUE)
  }
  
  files <- list.files(schema_dir, pattern = "\\.yaml$")
  log_msg(sprintf("  Found %d schema files", length(files)))
  
  # Check for key schemas
  expected <- c("CME.yaml", "EDUCATION.yaml", "NUTRITION.yaml")
  found <- sum(expected %in% files)
  log_msg(sprintf("  Key schemas present: %d/%d", found, length(expected)))
  
  return(length(files) >= 10 && found >= 2)
}

# ============================================================================
# NETWORK TESTS - Call UNICEF API (skipped in CI)
# These require sourcing the package code first
# ============================================================================

source_package_code <- function() {
  # Source package code only when needed (for network tests)
  if (file.exists(file.path(R_DIR, "get_unicef.R"))) {
    source(file.path(R_DIR, "get_unicef.R"))
    source(file.path(R_DIR, "metadata.R"))
    source(file.path(R_DIR, "flows.R"))
    source(file.path(R_DIR, "aliases_devtests.R"))
    return(TRUE)
  }
  return(FALSE)
}

test_list_flows_api <- function() {
  log_msg("Testing list_dataflows() from API...")
  
  if (!source_package_code()) {
    log_msg("  SKIP: Could not source package code")
    return(TRUE)
  }
  
  flows <- list_dataflows()
  
  log_msg(sprintf("  Found %d dataflows", nrow(flows)))
  
  write.csv(flows, file.path(OUTPUT_DIR, "test_dataflows.csv"), 
            row.names = FALSE, fileEncoding = "UTF-8")
  log_msg("  Saved to test_dataflows.csv")
  
  return(nrow(flows) > 50)
}

test_child_mortality_api <- function() {
  log_msg("Testing child mortality API (CME_MRY0T4)...")
  
  if (!exists("get_unicef", mode = "function")) {
    if (!source_package_code()) {
      log_msg("  SKIP: Could not source package code")
      return(TRUE)
    }
  }
  
  df <- get_unicef(
    indicator = "CME_MRY0T4",
    countries = c("USA", "GBR", "FRA"),
    start_year = 2020,
    end_year = 2023
  )
  
  log_msg(sprintf("  Retrieved %d observations", nrow(df)))
  
  if (!is.null(df) && nrow(df) > 0) {
    write.csv(df, file.path(OUTPUT_DIR, "test_mortality.csv"), row.names = FALSE)
    log_msg("  Saved to test_mortality.csv")
    return(nrow(df) > 0)
  }
  
  return(FALSE)
}

# ============================================================================
# Run All Tests
# ============================================================================

run_all_tests <- function() {
  cat("============================================================\n")
  cat("UNICEF API R Package Test Suite\n")
  cat(sprintf("Started: %s\n", Sys.time()))
  cat(sprintf("Environment: %s\n", if (IN_CI) "CI (GitHub Actions)" else "Local"))
  cat(sprintf("Network available: %s\n", if (NETWORK_AVAILABLE) "Yes" else sprintf("No (%s)", SKIP_REASON)))
  cat(sprintf("Metadata dir: %s (exists: %s)\n", METADATA_DIR, dir.exists(METADATA_DIR)))
  cat("============================================================\n\n")
  
  # OFFLINE tests - always run (fast, no network)
  # Only requires: yaml package (no sourcing of package code)
  cat("--- OFFLINE TESTS (bundled YAML metadata) ---\n\n")
  tests <- list(
    list(name = "YAML Dataflows", fn = test_yaml_dataflows),
    list(name = "YAML Indicators", fn = test_yaml_indicators),
    list(name = "YAML Countries", fn = test_yaml_countries),
    list(name = "Schema Files Exist", fn = test_schema_files_exist),
    list(name = "CME Schema Structure", fn = test_dataflow_schema_cme),
    list(name = "EDUCATION Schema Structure", fn = test_dataflow_schema_education)
  )
  
  # NETWORK tests - only run when network is available (not in CI, and connectivity confirmed)
  if (NETWORK_AVAILABLE) {
    cat("\n--- NETWORK TESTS (API calls) ---\n\n")
    tests <- c(tests, list(
      list(name = "List Dataflows API", fn = test_list_flows_api),
      list(name = "Child Mortality API", fn = test_child_mortality_api)
    ))
  } else {
    cat(sprintf("\n--- Skipping network tests (%s) ---\n\n", SKIP_REASON))
  }
  
  results <- list()
  
  for (test in tests) {
    result <- tryCatch({
      passed <- test$fn()
      list(name = test$name, status = if (passed) "PASS" else "FAIL", error = NULL)
    }, error = function(e) {
      log_msg(sprintf("  ERROR: %s", e$message))
      list(name = test$name, status = "ERROR", error = e$message)
    })
    results[[length(results) + 1]] <- result
    cat("\n")
  }
  
  cat("============================================================\n")
  cat("TEST RESULTS\n")
  cat("============================================================\n")
  
  passed_count <- 0
  for (r in results) {
    icon <- if (r$status == "PASS") "PASS" else "FAIL"
    cat(sprintf("[%s] %s\n", icon, r$name))
    if (!is.null(r$error)) {
      cat(sprintf("       Error: %s\n", r$error))
    }
    if (r$status == "PASS") passed_count <- passed_count + 1
  }
  
  cat(sprintf("\nTotal: %d/%d tests passed\n", passed_count, length(results)))
  cat("============================================================\n")
  
  # Exit with error if tests failed (for CI)
  if (passed_count < length(results) && IN_CI) {
    quit(status = 1)
  }
  
  invisible(passed_count == length(results))
}

# Run tests
run_all_tests()
